sdk/ml/azure-ai-ml/azure/ai/ml/_artifacts/_artifact_utilities.py

```diff
@@ -359,6 +370,125 @@ def _update_gen2_metadata(name, version, indicator_file, storage_client) -> None
     artifact_directory_client.set_metadata(_build_metadata_dict(name=name, version=version))
 
 
+def _generate_temporary_data_reference_id() -> str:
+    """Generate a temporary data reference id."""
+    return str(uuid.uuid4())
+
+
+@retry(
+    exceptions=HttpResponseError,
+    failure_msg="Artifact upload exceeded maximum retries. Try again.",
+    logger=module_logger,
+    max_attempts=MAX_ASSET_STORE_API_CALL_ATTEMPTS,
+)
+def _get_snapshot_temporary_data_reference(
+    asset_name: str,
+    asset_version: str,
+    request_headers: Dict[str, str],
+    workspace: Workspace,
+    requests_pipeline: HttpPipeline,
+    ) -> Tuple[str, str]:
+    """
+    Make a temporary data reference for an asset and return SAS uri and blob storage uri.
+
+    :param asset_name: Name of the asset to be created
+    :type asset_name: str
+    :param asset_version: Version of the asset to be created
+    :type asset_version: str
+    :param request_headers: Request headers for API call
+    :type request_headers: Dict[str, str]
+    :param workspace: Workspace object
+    :type workspace: azure.ai.ml._restclient.v2022_05_01.models.Workspace
+    :param requests_pipeline: Proxy for sending HTTP requests
+    :type requests_pipeline: azure.ai.ml._utils._http_utils.HttpPipeline
+    :return: Existing asset's name and version, if found
+    :rtype: Tuple[str, str]
+    """
+
+    # create temporary data reference
+    temporary_data_reference_id = _generate_temporary_data_reference_id()
+
+    # build and send request
+    asset_id = f"azureml://locations/{workspace.location}/workspaces/{workspace.workspace_id}/"\
+        f"codes/{asset_name}/versions/{asset_version}"
+    data = {
+        "assetId": asset_id,
+        "temporaryDataReferenceId": temporary_data_reference_id,
+        "temporaryDataReferenceType": "TemporaryBlobReference"
+    }
+    data_encoded = json.dumps(data).encode('utf-8')
+    serialized_data = json.loads(data_encoded)
+
+    # make sure correct cloud endpoint is used
+    cloud_endpoint = _get_cloud_details()['registry_discovery_endpoint']
+    service_url = replace_between(cloud_endpoint, HTTPS_PREFIX, ".", workspace.location)
+
+    # send request
+    request_url = f"{service_url}assetstore/v1.0/temporaryDataReference/createOrGet"
```

Comment:
Why there's not a service client for this? We have one for registry data reference. https://github.com/Azure/azure-sdk-for-python/blob/2fe6e3da6738994140b995584bfe1e980b192064/sdk/ml/azure-ai-ml/azure/ai/ml/_utils/_registry_utils.py#L77

Reply:
Agree! I would consider resuing the registry_util part for this if possible

Reply:
I mentioned in the comment above, but the registry temp data reference sas call is registry-specific. There isn't one yet for non-registry assets and the timeline for one would be spring at earliest which would leave Office blocked for months. 

While waiting for the MFE contract change, it was decided we would implement it directly here.